x/sdk/p2p/client.go

```diff
@@ -0,0 +1,184 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"golang.org/x/exp/maps"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/utils/sampler"
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+var (
+	ErrAppRequestFailed = errors.New("app request failed")
+	ErrRequestPending   = errors.New("request pending")
+)
+
+type AppResponseCallback func(
+	nodeID ids.NodeID,
+	responseBytes []byte,
+	err error,
+)
+type CrossChainAppResponseCallback func(
+	chainID ids.ID,
+	responseBytes []byte,
+	err error,
+)
+
+type Client struct {
+	handlerID uint8
+	router    *Router
+	sender    common.AppSender
+}
+
+func (c *Client) AppRequestAny(
+	ctx context.Context,
+	appRequestBytes []byte,
+	onResponse AppResponseCallback,
+) error {
+	c.router.lock.RLock()
+	peers := maps.Keys(c.router.peers)
+	c.router.lock.RUnlock()
+
+	s := sampler.NewUniform()
+	s.Initialize(uint64(len(peers)))
+	drawn, err := s.Next()
+	if err != nil {
+		return err
+	}
+
+	nodeIDs := set.Set[ids.NodeID]{peers[drawn]: struct{}{}}
+	return c.AppRequest(ctx, nodeIDs, appRequestBytes, onResponse)
+}
+
+func (c *Client) AppRequest(
+	ctx context.Context,
+	nodeIDs set.Set[ids.NodeID],
+	appRequestBytes []byte,
+	onResponse AppResponseCallback,
+) error {
+	appRequestBytes = c.prefixMessage(appRequestBytes)
+	for nodeID := range nodeIDs {
+
+		c.router.lock.Lock()
+		requestID := c.router.requestID
+
+		if _, ok := c.router.pendingAppRequests[requestID]; ok {
+			return fmt.Errorf(
+				"failed to issue request with request id %d: %w",
+				requestID,
+				ErrRequestPending,
+			)
+		}
+
+		if err := c.sender.SendAppRequest(
+			ctx,
+			set.Set[ids.NodeID]{nodeID: struct{}{}},
+			requestID,
+			appRequestBytes,
+		); err != nil {
+			return err
+		}
+
+		c.router.pendingAppRequests[requestID] = pendingAppRequest{
+			onResponse: onResponse,
+		}
+
+		c.router.requestID++
+		c.router.lock.Unlock()
```

Comment:
The locking is pretty weird here. Can we grab the lock once outside of the loop (and defer the release of the lock?).

There is currently a deadlock if and error is returned in the loop because the lock is never released.